src/Native/System.Native/pal_networking.cpp

```diff
@@ -287,15 +299,219 @@ extern "C" int32_t GetHostEntryForName(const uint8_t* address, HostEntry* entry)
     return PAL_EAI_SUCCESS;
 }
 
-extern "C" int32_t GetNextIPAddress(void** addressListHandle, IPAddress* endPoint)
+static int ConvertGetHostErrorPlatformToPal(int error)
 {
-    if (addressListHandle == nullptr || endPoint == nullptr)
+    switch (error)
     {
-        return PAL_EAI_BADARG;
+        case HOST_NOT_FOUND:
+            return PAL_HOST_NOT_FOUND;
+
+        case TRY_AGAIN:
+            return PAL_TRY_AGAIN;
+
+        case NO_RECOVERY:
+            return PAL_NO_RECOVERY;
+
+        case NO_DATA:
+            return PAL_NO_DATA;
+
+        default:
+            assert(false && "Unknown gethostbyname/gethostbyaddr error code");
+            return PAL_HOST_NOT_FOUND;
+    }
+}
+
+static void ConvertHostEntPlatformToPal(HostEntry& hostEntry, hostent& entry)
+{
+    hostEntry = {
+        .CanonicalName = reinterpret_cast<uint8_t*>(entry.h_name),
+        .Aliases = reinterpret_cast<uint8_t**>(entry.h_aliases),
+        .AddressListHandle = reinterpret_cast<void*>(&entry),
+        .IPAddressCount = 0,
+        .HandleType = HOST_ENTRY_HANDLE_HOSTENT
+    };
+
+    for (int i = 0; entry.h_addr_list[i] != nullptr; i++)
+    {
+        hostEntry.IPAddressCount++;
+    }
+}
+
+#if HAVE_GETHOSTBYNAME_R
+static int GetHostByNameHelper(const uint8_t* hostname, hostent** entry)
+{
+    assert(hostname != nullptr);
+    assert(entry != nullptr);
+
+    size_t scratchLen = 512;
+    uint8_t* buffer = reinterpret_cast<uint8_t*>(malloc(sizeof(hostent) + scratchLen));
+
+    hostent* result = reinterpret_cast<hostent*>(buffer);
+    char* scratch = reinterpret_cast<char*>(&buffer[sizeof(hostent)]);
+
+    for (;;)
+    {
+        int getHostErrno;
+        int err = gethostbyname_r(reinterpret_cast<const char*>(hostname), result, scratch, scratchLen, entry, &getHostErrno);
+        switch (err)
+        {
+            case 0:
+                *entry = result;
+                return 0;
+
+            case ERANGE:
+                free(buffer);
+
+                scratchLen *= 2;
+                buffer = reinterpret_cast<uint8_t*>(malloc(sizeof(hostent) + scratchLen));
+                if (buffer == nullptr)
+                {
+                    return PAL_NO_MEM;
+                }
+
+                result = reinterpret_cast<hostent*>(buffer);
+                scratch = reinterpret_cast<char*>(&buffer[sizeof(hostent)]);
```

Comment:
If you were to move everything after the scratchLen *= 2 to the top of the for loop, it looks like you could consolidate both memory accesses and region calculations into a single location, just having this free the buffer, double scratchLen, and continue.